include/revng/Pipeline/DynamicHierarchy.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <iostream>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/raw_ostream.h"
+
+/// a dynamic hierarcy is a tree that can be extended by
+/// downstream libraries. By instantiating a node and providing a
+/// parent then the child will be visible from the parent as well.
+///
+/// notice that init must be called after every object of the hierarchy has been
+/// constructed.
+template<typename DerivedType>
+class DynamicHierarchy {
+public:
+  using entry_t = size_t;
+
+  DynamicHierarchy(llvm::StringRef Name) : Name(Name.str()) {
+    DerivedType::Roots.push_back(&self());
```

Comment:
Why use a private field of the derived class as opposed to put it in the base class?